clang/include/clang/Basic/AttrDocs.td

```diff
@@ -3241,6 +3241,73 @@ parameter.
   }];
 }
 
+def SYCLIntelFPGAIpInterfaceCsrDocs : Documentation {
+  let Category = DocCatFunction;
+  let Heading = "intel::ip_interface_csr";
+  let Content = [{
+``[[intel::ip_interface_csr]]`` attribute applies to a function/lambda function.
+Takes no arguments and applies to the IP component that uses CSR
+(Control-Status Registers) invocation interface. Cannot be applied multiple times
+to the same function. Cannot be combined with``[[intel::ip_interface_streaming]]``
+or ``[[intel::ip_stall_free_return]]``.
+
+if ``intel::ip_interface_csr`` is applied to a function called from a
+device kernel, the attribute is ignored and it is not propagated to the kernel.
+
+.. code-block:: c++
+
+  [[intel::ip_interface_csr]] int foo() {}
+
+  }];
+}
+
+def SYCLIntelFPGAIpInterfaceStreamingDocs : Documentation {
+  let Category = DocCatFunction;
+  let Heading = "intel::ip_interface_streaming";
+  let Content = [{
+``[[intel::ip_interface_streaming]]`` attribute applies to a function/lambda function.
+Takes no arguments and applies to the IP component that supports streaming invocation
+interface via a handshaking protocol. Cannot be applied multiple times to the same
+function. Cannot be combined with ``[[intel::ip_interface_csr]]`` but can be used with
+``[[intel::ip_stall_free_return]]`` to the same function.
+
+if ``intel::ip_interface_streaming`` is applied to a function called from a
+device kernel, the attribute is ignored and it is not propagated to the kernel.
+
+.. code-block:: c++
+
+  [[intel::ip_interface_streaming]] int foo() {}
+
+  [[intel::ip_interface_streaming]]
+  [[intel::ip_stall_free_return]] int foo1() {}
+
+  }];
+}
+
+def SYCLIntelFPGAIpStallFreeReturnDocs : Documentation {
+  let Category = DocCatFunction;
+  let Heading = "intel::ip_stall_free_return";
+  let Content = [{
+``[[intel::ip_stall_free_return]]`` attribute applies to a function/lambda function.
+Takes no arguments and applies to IP components with a streaming invocation interface.
+This is an optimization hint that the return interface of the IP component will never
+be stalled from downstream. Cannot be applied multiple times to the same
```

Comment:
This one is a little bit better in that I at least know it's an optimization hint. How would a user know when to write this attribute, though? I don't know what it means to "never be stalled from downstream", but what's the effect if the user writes this in a situation where the function actually is stalled from downstream?